### ISSUE TITLE
Finansial analisis

### ISSUE THREAD

**Issue by am5668260-del** [tyron12233/CodeAssist#1600]
import yfinance as yf
import pandas as pd
import matplotlib.pyplot as plt

def get_stock_data(ticker, period="6mo"):
    """Mengambil data historis saham dari Yahoo Finance."""
    print(f"Mengambil data saham: {ticker}...")
    stock = yf.Ticker(ticker)
    df = stock.history(period=period)
    return df

def calculate_indicators(df):
    """Menghitung indikator Moving Average (MA20, MA50) dan RSI."""
    # Moving Average
    df['MA20'] = df['Close'].rolling(window=20).mean()
    df['MA50'] = df['Close'].rolling(window=50).mean()

    # Relative Strength Index (RSI - 14 periode)
    delta = df['Close'].diff()
    gain = (delta.where(delta > 0, 0)).rolling(window=14).mean()
    loss = (-delta.where(delta < 0, 0)).rolling(window=14).mean()
    rs = gain / loss
    df['RSI'] = 100 - (100 / (1 + rs))

    return df

def generate_analysis(df, ticker):
    """Memberikan analisis otomatis berdasarkan data terbaru."""
    latest = df.iloc[-1]
    prev = df.iloc[-2]

    print("\n" + "="*40)
    print(f"   RINGKASAN ANALISIS SAHAM: {ticker.upper()}")
    print("="*40)
    print(f"Harga Penutupan Terakhir : ${latest['Close']:.2f}")
    print(f"MA20                     : ${latest['MA20']:.2f}")
    print(f"MA50                     : ${latest['MA50']:.2f}")
    print(f"RSI (14)                 : {latest['RSI']:.2f}")
    print("-" * 40)
    print("Sinyal & Analisis:")

    # Analisis Trend (Moving Average)
    if latest['Close'] > latest['MA20'] > latest['MA50']:
        print("- Trend: UPTREND Kuat (Harga di atas MA20 & MA50).")
    elif latest['Close'] < latest['MA20'] < latest['MA50']:
        print("- Trend: DOWNTREND Kuat (Harga di bawah MA20 & MA50).")
    else:
        print("- Trend: SIDEWAYS / Konsolidasi.")

    # Analisis RSI
    if latest['RSI'] >= 70:
        print("- RSI: OVERBOUGHT (Jenuh Beli) -> Waspada potensi koreksi/penurunan harga.")
    elif latest['RSI'] <= 30:
        print("- RSI: OVERSOLD (Jenuh Jual) -> Waspada potensi pantulan/kenaikan harga.")
    else:
        print("- RSI: NETRAL (Kondisi stabil).")

    # Sinyal Golden/Death Cross Sederhana
    if prev['MA20'] < prev['MA50'] and latest['MA20'] > latest['MA50']:
        print("- Sinyal Khusus: GOLDEN CROSS Terdeteksi (Sinyal Beli Potensial)!")
    elif prev['MA20'] > prev['MA50'] and latest['MA20'] < latest['MA50']:
        print("- Sinyal Khusus: DEATH CROSS Terdeteksi (Sinyal Jual Potensial)!")

    print("="*40 + "\n")

def plot_chart(df, ticker):
    """Menampilkan grafik harga saham dan RSI."""
    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(12, 8), gridspec_kw={'height_ratios': [3, 1]})

    # Plot Grafik Harga
    ax1.set_title(f"Grafik Analisis Saham {ticker.upper()}", fontsize=14)
    ax1.plot(df.index, df['Close'], label='Harga Closing', color='blue', alpha=0.6)
    ax1.plot(df.index, df['MA20'], label='MA 20', color='orange', linestyle='--')
    ax1.plot(df.index, df['MA50'], label='MA 50', color='red', linestyle='--')
    ax1.set_ylabel("Harga")
    ax1.legend()
    ax1.grid(True)

    # Plot Grafik RSI
    ax2.plot(df.index, df['RSI'], label='RSI 14', color='purple')
    ax2.axhline(70, color='red', linestyle='--', alpha=0.7, label='Overbought (70)')
    ax2.axhline(30, color='green', linestyle='--', alpha=0.7, label='Oversold (30)')
    ax2.set_ylabel("RSI")
    ax2.set_ylim(0, 100)
    ax2.legend(loc='lower left')
    ax2.grid(True)

    plt.tight_layout()
    plt.show()

def main():
    print("--- APLIKASI MEMBACA ANALISIS SAHAM ---")
    ticker_input = input("Masukkan Ticker Saham (contoh: BBCA.JK untuk Indo, AAPL untuk US): ").strip()
    
    if not ticker_input:
        ticker_input = "AAPL"  # Default jika dikosongkan

    try:
        # 1. Fetch Data
        df = get_stock_data(ticker_input, period="6mo")
        
        if df.empty:
            print("Data tidak ditemukan! Pastikan simbol/ticker benar.")
            return

        # 2. Hitung Indikator
        df = calculate_indicators(df)

        # 3. Tampilkan Hasil Analisis Teks
        generate_analysis(df, ticker_input)

        # 4. Tampilkan Grafik
        plot_chart(df, ticker_input)

    except Exception as e:
        print(f"Terjadi kesalahan: {e}")

if __name__ == "__main__":
    main()
